modin/engines/base/frame/data.py

```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
             all(o.axes[axis].equals(self.axes[axis]) for o in other)
             and not is_aligning_applied
         ):
+            # aligning self.list_of_blocks
             return (
                 self._partitions,
                 [self._simple_shuffle(axis, o) for o in other],
                 self.axes[axis].copy(),
             )
+
         index_other_obj = [o.axes[axis] for o in other]
         joined_index = self._join_index_objects(axis, index_other_obj, how, sort)
         # We have to set these because otherwise when we perform the functions it may
         # end up serializing this entire object.
         left_old_idx = self.axes[axis]
         right_old_idxes = index_other_obj
 
-        is_avoid_reindex = len(joined_index) != len(joined_index.unique()) and axis == 0
+        def make_map_func():
+            if len(joined_index) != len(joined_index.unique()) and axis == 0:
+                return lambda df: df
+            return lambda df: df.reindex(joined_index, axis=axis)
+
         # Start with this and we'll repartition the first time, and then not again.
-        if (
-            not is_aligning_applied
-            and not is_avoid_reindex
-            and (force_repartition or not left_old_idx.equals(joined_index))
+        if is_aligning_applied or (
+            not force_repartition and left_old_idx.equals(joined_index)
         ):
+            reindexed_self = self._partitions
+        else:
+            # aligning index without aligning partition' blocks
```

Comment:
This phrase is unclear as well, as for me, as partition' blocks are reshuffled as well. On the whole, I would either delete new comments at all or write more descriptive comments.